.github/workflows/tox.yml

```diff
@@ -229,3 +229,43 @@ jobs:
         uses: codecov/codecov-action@v4
         with:
           token: ${{ secrets.CODECOV_TOKEN }}
+
+      - if: ${{ (success() || failure()) && contains(matrix.coverage, 'github') && matrix.pytest == 'true' }}
+        run: |
+          mv .coverage .coverage.${{ runner.os }}-${{ runner.arch }}-${{ matrix.toxenv }}
+          ls -la .
+        shell: sh
+
+      - name: Upload coverage data to GitHub
+        if: ${{ (success() || failure()) && contains(matrix.coverage, 'github') && matrix.pytest == 'true' }}
+        uses: actions/upload-artifact@v4
+        with:
+          name: .coverage.${{ runner.os }}-${{ runner.arch }}-${{ matrix.toxenv }}
+          path: .coverage.${{ runner.os }}-${{ runner.arch }}-${{ matrix.toxenv }}
+
+  coverage:
+    name: report coverage
+    needs:
+      - tox
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+          lfs: true
+          submodules: ${{ inputs.submodules }}
+          ref: ${{ inputs.checkout_ref }}
+      - uses: actions/download-artifact@v4
+        with:
+          pattern: .coverage.*
+          merge-multiple: true
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.12"
+        continue-on-error: true
```

Comment:
I don't know how I feel on so many usage of `continue-on-error: true` in this workflow. I would want to know if a step errored. And I'd rather a step not run when it is not supposed to run using `if` instead of throwing error.

Reply:
Otherwise, when a real error happens, it is hard to debug because everything is green no matter what.

Reply:
that is fair; the ideal solution would be to run this job conditionally (`if: `) only when `matrix.coverage` contains `github`; then we could remove these `continue-on-error`s.

However, I'm not sure how to actually do that. I tried doing `if: inputs.coverage == 'github'`, but `inputs.coverage` to this workflow was always blank and did not reflect the contents of the generated matrix.

Reply:
Where did you put the `if`? We can put this at the job level, right, since this whole "report coverage" is irrelevant for other modes?

For the combining part, maybe can skip if no coverage files found after download (maybe can use `hashFiles`, see https://stackoverflow.com/questions/71336204/github-action-check-if-a-file-already-exists), and maybe the last 3 Python commands can be just one step.

Reply:
oh, good point! I put the `if` at the job level, but we could put it in the individual steps just as easily